[0.80.0][kotlin] fixed breaking changes in 0.80 after kotlin conversion

chrfalch opened the pull request:
# Why

React Native 0.80 introduces some breaking changes after converting from Java -> Kotlin. Mainly about changing nullability on parameter types. This is intended and should not be fixed in RN.

# How

Fixed our use of the APIs.

# Test Plan

CI

# Checklist

- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

chrfalch committed:
[sdk-53][kotlin] fixed breaking changes in 0.80 after kotlin conversion

After Kotlin conversion the Dynamic namespace had a breaking change where all conversions are now optional.

- This commit fixes this by casting a DynamicCastException when a method returns null
- Implemented DynamicCastException which inherits from CodedException
- Fixed invalid shadowing of property name in KModuleEventEmitterWrapper.kt

chrfalch commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/expo/expo/37039?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#37039** <a href="https://app.graphite.dev/github/pr/expo/expo/37039?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/expo/expo/37039?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#37034** <a href="https://app.graphite.dev/github/pr/expo/expo/37034?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#37033** <a href="https://app.graphite.dev/github/pr/expo/expo/37033?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#36932** <a href="https://app.graphite.dev/github/pr/expo/expo/36932?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

chrfalch committed:
Fixed breaking java -> kotlin changes after rebase

After rebase with expo's main this commit fixes a few more type issues we're seeing

chrfalch committed:
Fixed breaking java -> kotlin changes

After rebase UriTypeConverter expected String?. This commit fixes this.

Kudo commented on packages/expo-image/android/src/main/java/expo/modules/image/okhttp/GlideUrlWrapperLoader.kt:
```diff
@@ -28,7 +28,7 @@ class GlideUrlWrapperLoader(
               .newBuilder()
               .body(
-                ProgressResponseBody(originalResponse.body) { bytesWritten, contentLength, done ->
+                ProgressResponseBody(originalResponse.body ?: throw NullPointerException()) { bytesWritten, contentLength, done ->
```
```suggestion
                ProgressResponseBody(requireNotNull(originalResponse.body)) { bytesWritten, contentLength, done ->
```

nit: NullPointerException is not quite precise and we can just use requireNotNull

Kudo commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/exception/CodedException.kt:
```diff
@@ -247,6 +247,13 @@ internal class CollectionElementCastException private constructor(
 
+@DoNotStrip
+class DynamicCastException(
+  typeName: String
```
maybe we can pass class type and get its string name rather than having the handy `map`, `string`, `array` as inputs

chrfalch replied:
Agree. Have changed this now!

chrfalch committed:
Update GlideUrlWrapperLoader.kt

chrfalch committed:
Updated DynamicCastException to take type parameter

Instead of taking a string as its parameter it now takes a type and infers the name from it.
